<a href="https://colab.research.google.com/github/rmehdi1/CommunityProject_Mobilize/blob/main/ChangeOrgAnalysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Import Libraries

In [10]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import kagglehub
import pandas as pd
import os

# Import Data

In [27]:
# Download latest version of the dataset
path = kagglehub.dataset_download("muhammedabdulazeem/petitions-from-changeorg")
print("Path to dataset files:", path)

# Define dataset path
dataset_path = "/root/.cache/kagglehub/datasets/muhammedabdulazeem/petitions-from-changeorg/versions/1"

# Load the CSV file into a DataFrame
df = pd.read_csv(os.path.join(dataset_path, "results.csv"))

Path to dataset files: /kaggle/input/petitions-from-changeorg


/tmp/ipython-input-27-1786828998.py:9: DtypeWarning: Columns (14) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(os.path.join(dataset_path, "results.csv"))


**Dataset Details: Petitions from Change.org**

This project uses the **“Petitions from Change.org”** dataset from Kaggle, which contains over **24,000 public petitions** collected from [Change.org](https://www.change.org) using web scraping techniques (via the Python library `scrapy`). It contains petions Petitions span a wide range of social, environmental, and political issues.


**Why Change.Org:**

Change.org is the world’s largest online petition platform, making it an ideal source for studying grassroots campaign messaging. Its rich success metrics (like signatures and victories) offer measurable insights into what makes advocacy messaging effective.

See More about it at:
https://www.change.org/about

https://www.linkedin.com/company/change-org/




# EDA

In [19]:
df.head()

Unnamed: 0  petition_id                                       petition_ask  \
0           0     28127009               Free Kusha and reunite elephant pair   
1           1     28569017                          Save Caracal cat in india   
2           2     28566320  Animal Treatment and ABC Centre for Borivli, K...   
3           3     28543362  Seven Sebastian to say the N word once Unbanne...   
4           4     28471106                     Don't buy animals, Adopt them!   

             created_at  deleted_at original_locale  \
0  2021-04-01T00:07:47Z         NaN           en-IN   
1  2021-04-25T19:56:09Z         NaN           en-IN   
2  2021-04-25T15:40:17Z         NaN           en-IN   
3  2021-04-24T06:29:05Z         NaN           en-IN   
4  2021-04-20T03:45:00Z         NaN           en-IN   

                                               title  \
0  Karnataka Forest Department: Free Kusha and re...   
1  Save Caracal cat in India: @BoardWelfare, @Man...   
2  Municipal Commissioner of Mumbai: Animal Treat...   
3  Facebook: Seven Sebastian to say the N word on...   
4  People for the Ethical Treatment of Animals (P...   

                                         description  \
0  <p>Kusha, a male elephant from Dubare Elephant...   
1  <p>#savefloppainindia trend and sharing this p...   
2  <p>In Kandivli, Borivli Dahisar (Ward R), lack...   
3  <p>SVen sebaAstian was banned on Fb and that w...   
4  <p>I feel that animals are not just a thing to...   

                               targeting_description  \
0  karnataka forest department, People for the Et...   
1  Shri Prakash Javedkar (Cabinet Minister - Mini...   
2                   Municipal Commissioner of Mumbai   
3                                  Facebook, Germany   
4  People for the Ethical Treatment of Animals (P...   

                                         letter_body  ... total_signup_count  \
0               Free Kusha and reunite elephant pair  ...                 84   
1                          Save Caracal cat in india  ...                  8   
2  Animal Treatment and ABC Centre for Borivli, K...  ...                108   
3  Seven Sebastian to say the N word once Unbanne...  ...                  2   
4                     Don't buy animals, Adopt them!  ...                 46   

   supporter_message_count hide_comments hide_dm_action_panel  \
0                      NaN         False                False   
1                      NaN         False                False   
2                      NaN         False                False   
3                      NaN         False                False   
4                      NaN         False                False   

  enable_human_verification  \
0                     False   
1                     False   
2                     False   
3                     False   
4                     False   

                                       display_title  \
0  Karnataka Forest Department: Free Kusha and re...   
1  Save Caracal cat in India: @BoardWelfare, @Man...   
2  Municipal Commissioner of Mumbai: Animal Treat...   
3  Facebook: Seven Sebastian to say the N word on...   
4  People for the Ethical Treatment of Animals (P...   

                                      petition_title  petition_status  \
0  Karnataka Forest Department: Free Kusha and re...           active   
1  Save Caracal cat in India: @BoardWelfare, @Man...           active   
2  Municipal Commissioner of Mumbai: Animal Treat...           active   
3  Facebook: Seven Sebastian to say the N word on...           active   
4  People for the Ethical Treatment of Animals (P...           active   

   calculated_goal   progress  
0             7500  88.106667  
1             5000  52.920000  
2             2500  94.680000  
3              100   6.000000  
4              500  97.200000  

[5 rows x 41 columns]

In [21]:
df.describe()


Unnamed: 0   petition_id  deleted_at          goal           lat  \
count  24065.00000  2.406500e+04         0.0     25.000000  21844.000000   
mean   12032.00000  2.512415e+07         NaN  24988.000000     20.658804   
std     6947.11145  3.898197e+06         NaN  19856.239993      3.031797   
min        0.00000  3.433500e+04         NaN    700.000000    -29.532800   
25%     6016.00000  2.315384e+07         NaN   3000.000000     20.593700   
50%    12032.00000  2.630225e+07         NaN  25000.000000     20.593700   
75%    18048.00000  2.806185e+07         NaN  50000.000000     20.593700   
max    24064.00000  2.867789e+07         NaN  50000.000000     55.378100   

               long  displayed_signature_count  total_signature_count  \
count  21844.000000               2.406500e+04           2.406500e+04   
mean      78.641688               8.002824e+03           7.006357e+03   
std        8.266192               5.252928e+04           4.335863e+04   
min     -122.000000               5.000000e+00           5.000000e+00   
25%       78.962900               1.500000e+01           1.500000e+01   
50%       78.962900               8.000000e+01           8.000000e+01   
75%       78.962900               6.410000e+02           6.410000e+02   
max      145.491000               1.004861e+06           1.002221e+06   

       weekly_signature_count  total_page_views  total_share_count  \
count            24065.000000      2.406500e+04       2.406500e+04   
mean               236.993601      1.038450e+04       2.573867e+03   
std               2373.335908      1.096157e+05       2.794019e+04   
min                  0.000000      0.000000e+00       0.000000e+00   
25%                  0.000000      8.000000e+01       1.200000e+01   
50%                  0.000000      2.740000e+02       3.800000e+01   
75%                  1.000000      1.288000e+03       2.120000e+02   
max              55762.000000      3.707960e+06       1.015223e+06   

       daily_signature_count  monthly_signature_count  daily_signup_count  \
count           24065.000000             24065.000000        24065.000000   
mean               58.250031               491.874257            0.626512   
std               859.720190              3582.964189            7.928881   
min                 0.000000                 0.000000            0.000000   
25%                 0.000000                 0.000000            0.000000   
50%                 0.000000                 1.000000            0.000000   
75%                 0.000000                16.000000            0.000000   
max             21713.000000             75298.000000          228.000000   

       weekly_signup_count  monthly_signup_count  total_signup_count  \
count         24065.000000          24065.000000        24065.000000   
mean             13.962394             51.625390         1164.662373   
std             265.901304            586.680897        14908.715998   
min               0.000000              0.000000            0.000000   
25%               0.000000              0.000000            5.000000   
50%               0.000000              0.000000           16.000000   
75%               0.000000              2.000000           96.000000   
max           13861.000000          22606.000000       571196.000000   

       supporter_message_count  calculated_goal      progress  
count                      0.0     2.406500e+04  24065.000000  
mean                       NaN     1.075256e+04     48.226989  
std                        NaN     7.203991e+04     30.699688  
min                        NaN     1.000000e+02      5.000000  
25%                        NaN     1.000000e+02     15.000000  
50%                        NaN     1.000000e+02     52.300000  
75%                        NaN     1.000000e+03     76.026667  
max                        NaN     1.500000e+06     99.933333

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24065 entries, 0 to 24064
Data columns (total 41 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Unnamed: 0                 24065 non-null  int64  
 1   petition_id                24065 non-null  int64  
 2   petition_ask               24065 non-null  object 
 3   created_at                 24065 non-null  object 
 4   deleted_at                 0 non-null      float64
 5   original_locale            24065 non-null  object 
 6   title                      24065 non-null  object 
 7   description                24065 non-null  object 
 8   targeting_description      24057 non-null  object 
 9   letter_body                24065 non-null  object 
 10  end_date                   12816 non-null  object 
 11  goal                       25 non-null     float64
 12  slug                       24065 non-null  object 
 13  victory_date               970 non-null    obj

## **Null Value Exploration**

In [29]:
# Total number of rows
total_rows = len(df)

# Count and percent of missing values
null_counts = df.isnull().sum()
null_percent = (null_counts / total_rows) * 100

# Combine into a summary DataFrame
missing_summary = pd.DataFrame({
    'Null Count': null_counts,
    'Percent Missing': null_percent
})

# Filter only columns with missing values
missing_summary = missing_summary[missing_summary['Null Count'] > 0]

# Sort by % missing, descending
missing_summary = missing_summary.sort_values(by='Percent Missing', ascending=False)

# Display the result
print(missing_summary)


                         Null Count  Percent Missing
deleted_at                    24065       100.000000
supporter_message_count       24065       100.000000
victory_description           24064        99.995845
goal                          24040        99.896115
victory_date                  23095        95.969250
end_date                      11249        46.744234
lat                            2221         9.229171
long                           2221         9.229171
targeting_description             8         0.033243


### **Deleted Variables**

Removed columns with high missingness and minimal analytical value:
- `deleted_at` (100% missing)
- `supporter_message_count` (100% missing)
- `victory_description` (99.9% missing)

### **Variable: `goal` / `calculated_goal`**

The `goal` column is sparsely populated and likely deprecated. The `calculated_goal` column is complete, but based on Change.org’s public documentation and the dataset source, it may have been inferred by the scraper rather than sourced directly from the platform. Since Change.org dynamically adjusts visible goals to encourage engagement, we treat `calculated_goal` as an approximate milestone—not a definitive success target.


In [34]:
# Drop columns with excessive missing values or limited analytical use
columns_to_drop = [
    'deleted_at',
    'supporter_message_count',
    'victory_description',
    'goal'  # Sparse and likely deprecated
]

df.drop(columns=columns_to_drop, inplace=True)


**Exploring variable: `victory_date`**

In [35]:
# Ensure datetime format
df['victory_date'] = pd.to_datetime(df['victory_date'], errors='coerce')

# 1. Check how many total victories exist
total_victories = df['is_victory'].sum()

# 2. Count of rows where `is_victory` is True but `victory_date` is missing
victory_missing_date = df[(df['is_victory'] == True) & (df['victory_date'].isnull())].shape[0]

# 3. Count of rows where `is_victory` is False but `victory_date` is present (shouldn't happen)
non_victory_with_date = df[(df['is_victory'] == False) & (df['victory_date'].notnull())].shape[0]

# 4. Summary of non-null victory dates
victory_date_summary = df[df['victory_date'].notnull()]['victory_date'].describe()

# 5. Optional: create flag for analysis
df['has_victory_date'] = df['victory_date'].notnull()

# Print findings
print(f"Total victories: {total_victories}")
print(f"Victories missing `victory_date`: {victory_missing_date}")
print(f"Non-victories with `victory_date`: {non_victory_with_date}")
print("\nVictory date distribution (where not null):")
print(victory_date_summary)


Total victories: 938
Victories missing `victory_date`: 0
Non-victories with `victory_date`: 32

Victory date distribution (where not null):
count                              970
mean     2020-09-18 11:39:12.989690624
min                2017-03-12 00:00:00
25%                2020-06-21 12:00:00
50%                2021-01-13 00:00:00
75%                2021-04-16 00:00:00
max                2021-05-01 00:00:00
Name: victory_date, dtype: object




The `victory_date` field is valid only for petitions marked as victories. Remove inconsistent values where non-victorious petitions have a `victory_date`, then create a new feature to calculate the number of days it took to reach victory.


In [39]:
# Ensure both are datetime
df['created_at'] = pd.to_datetime(df['created_at'], errors='coerce')
df['victory_date'] = pd.to_datetime(df['victory_date'], errors='coerce')

# Strip timezone to make both tz-naive
df['created_at'] = df['created_at'].dt.tz_localize(None)
df['victory_date'] = df['victory_date'].dt.tz_localize(None)

# Now safely calculate days to victory
df['days_to_victory'] = (df['victory_date'] - df['created_at']).dt.days

# Nullify invalid victory dates
df.loc[df['is_victory'] == False, 'victory_date'] = pd.NaT

# Create duration feature (days from creation to victory)
df['days_to_victory'] = (df['victory_date'] - df['created_at']).dt.days

# keep flag
df['has_victory_date'] = df['victory_date'].notnull()

**Handling `end_date`**

In [40]:
# Ensure end_date is datetime
df['end_date'] = pd.to_datetime(df['end_date'], errors='coerce')

# 1. How many are missing?
missing_end = df['end_date'].isnull().sum()
print(f"Missing end_date: {missing_end} out of {len(df)}")

# 2. Check relationship with petition_status
print("\nMissing end_date by petition_status:")
print(df[df['end_date'].isnull()]['petition_status'].value_counts())

print("\nNon-missing end_date by petition_status:")
print(df[df['end_date'].notnull()]['petition_status'].value_counts())

# 3. Check relationship with is_victory
print("\nMissing end_date by is_victory:")
print(df[df['end_date'].isnull()]['is_victory'].value_counts())

print("\nNon-missing end_date by is_victory:")
print(df[df['end_date'].notnull()]['is_victory'].value_counts())

# 4. Create helper flag for missing end date
df['has_end_date'] = df['end_date'].notnull()


Missing end_date: 11249 out of 24065

Missing end_date by petition_status:
petition_status
active     10758
victory      491
Name: count, dtype: int64

Non-missing end_date by petition_status:
petition_status
active     12357
victory      447
closed        12
Name: count, dtype: int64

Missing end_date by is_victory:
is_victory
False    10758
True       491
Name: count, dtype: int64

Non-missing end_date by is_victory:
is_victory
False    12369
True       447
Name: count, dtype: int64




The `end_date` field indicates when a petition was closed. Nearly half the petitions have a missing `end_date`, mostly those still marked as active. We create an `is_active` flag and calculate the petition duration only for those with a known end date.
